# nanofmdv
This is a bioinformatics pipeline to process the Foot-and-Mouth Disease Virus (FMDV) amplicons generated by the Shaw et al. approach. The overall aim is to select an appropriate reference sequence from the FMDV reference sequence set, align the reads, trim the amplicon primers, and create a consensus sequence.

![NanoFMDV piepline](https://github.com/rjorton/nanofmdv/blob/main/nanofmdv-pipeline.png)

The bioinformatics pipeline is currently implemented via a collection of bash scripts, which call existing bioinformatics tools as well as custom Python scripts. The pipeline, along with the reference FMDV sequences, and usage instructions is publicly available at: https://github.com/rjorton/nanofmdv. 

The first step in the bioinformatics pipeline is to determine an appropriate FMDV reference sequence to use for read alignment. A database of 79 near complete FMDV genomes were downloaded from the FMDbase sequence database (https://www.fmdbase.org), representing approximately one per existing FMDV lineage; typically the FMDbase ‘RefSeq’ was selected for each lineage if available, unless a significantly longer genome sequence (ignoring N bases) was available. The VP1 sequence of each genome was extracted to create an equivalent FMDV-VP1 sequence data set.  

Each demultiplexed sample is analysed independently. After concatenating all a sample’s FASTQ reads into a single file, minimap2 is used to align the reads to the FMDV-VP1 sequence collection. The VP1 reference sequence with the highest number of reads mapping to it is selected as the reference sequence for the sample, with minimap2 again used to align all the reads on to the reference FMDV genome corresponding to the selected VP1 sequence. 

As tiled amplicons are used, amplicon primer sequences must be trimmed before consensus sequence generation. A simple Python script, based on the previous GoPrime methodology, is used to compare all the possible forward and reverse primers of each amplicon to the selected reference sequence to identify possible binding sites. A BED file is subsequently created representing the binding regions of each amplicon’s primers on the reference sequence. The samtools  ampliconclip function is then used to clip primer sequences from the aligned reads within the BAM file using the previously created BED file, with a primer-clipped read FASTQ file also generated using the samtools fastq function. 

A preliminary consensus sequence is generated using the samtools consensus function. This preliminary consensus sequence, along with the primer-clipped read FASTQ file, is then used by the nanopore variant/consensus caller medaka (https://github.com/nanoporetech/medaka) to call a ‘polished’ FMDV consensus genome sequence (medaka has models to address nanopore associated errors such as those at homopolymers), the consensus sequence is also depth masked to mask positions with read coverage less than 10 with N bases. 

A final step in the pipeline, is to BLASTn the final consenseus sequence against a collection of 923 FMDV genome sequences to determine the closest sequence matches and their respective serotype/topotype. Assembly statistics were generated using samtools.
